src/miner.cpp

```diff
@@ -1011,28 +1031,29 @@ void BitcoinMiner(std::shared_ptr<CReserveScript> coinbaseScript, bool fProofOfS
                     ++pblock->nNonce;
                 }
             } else {
-                LogPrintf("%s: Unknown hashing algorithm found!\n");
+                LogPrintf("%s: Unknown hashing algorithm found!\n", __func__);
                 return;
             }
 
             double nHashSpeed = 0;
             {
                 LOCK(cs_nonce);
-                nHashes += nTries;
+                nHashes += nTries + (nMidTries * nInnerLoopCount);
                 nTimeDuration = GetTime() - nTimeStart;
                 if (!nTimeDuration) nTimeDuration = 1;
                 nHashSpeed = arith_uint256(nHashes/1000/nTimeDuration).getdouble();
             }
-            LogPrint(BCLog::BLOCKCREATION, "%s: PoW Hashspeed %d kh/s\n", __func__, nHashSpeed);
 
-            if (nTries == nInnerLoopCount) {
+            LogPrint(BCLog::MINING, "%s: PoW Hashspeed %d kh/s\n", __func__,  nHashSpeed);
+            if (nMidTries == nMidLoopCount || fBlockFoundAlready) {
+                fBlockFoundAlready = false;
                 continue;
```

Comment:
This is a generic section for all mining algos, not just Sha256.  So while it will work for that, it will have problems with the others since they don't have `fBlockFoundAlready` or an incrementing `nMidTries`.  In other words, ProgPoW would miss this check since `nTries == nInnerLoopCount` but `fBlockFoundAlready` is false, and `nMidTries` is 0; so it would assume it found a block every pass.

This makes for one NACK

Reply:
This was straightforward to fix by adding another case to the conditional, since `nTries == nInnerLoopCount` is never true for SHA with this change.

Reply:
Fixed more simply by casing only on "success".